storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
 	})
 }
 
+type SelectRecord struct {
+	Record json.RawMessage
+	Err    error
+}
+
+func (s *S3) selectStream(resp *s3.SelectObjectContentOutput, resultch chan<- *SelectRecord) {
+	defer resp.EventStream.Close()
+
+	reader, writer := io.Pipe()
+	defer reader.Close()
+	go func() {
+		defer writer.Close()
+		// TODO: wire ctx up to eventStream.StreamCloser.Close
+		for event := range resp.EventStream.Reader.Events() {
+			switch e := event.(type) {
+			case *s3.RecordsEvent:
+				// TODO: how should we handle errors here?!
+				writer.Write(e.Payload)
+			// How the heck are these different?
+			case *s3.StatsEvent:
+			case *s3.ProgressEvent:
+			}
+		}
+	}()
+
+	decoder := json.NewDecoder(reader)
+	// decoder.DontCopyRawMessage()
```

Comment:
I tried doing some zerocopy stuff, but it looks like there's an internal buffer that gets reused, and doing that plus the delay between writing to the channel and reading from the channel causes data corruption.

Reply:
Let's do the optimization after this PR, if necessary of course.